swim/heal_partition_test.go

```diff
@@ -0,0 +1,343 @@
+	defer destroyNodes(B...)
+
+	A.AddPartitionWithStatus(B, Faulty)
+	B.AddPartitionWithStatus(A, Faulty)
+
+	A.ProgressTime(time.Millisecond)
+	B.ProgressTime(time.Millisecond)
+
+	A[0].node.discoverProvider = statichosts.New(append(A.Hosts(), B.Hosts()...)...)
+
+	A[0].node.healer.Start()
+
+	// Progress time in a background thread. This will cause the node to
+	// attempt a heal periodically.
+	c := clock.NewMock()
+	A[0].node.clock = c
+	go func() {
+		for {
+			c.Add(time.Second)
+		}
+	}()
+
+	waitForPartitionHeal(t, 3000*time.Millisecond, A, B)
+}
+
+// TestHealBeforeBootstrap tests that we can heal with a node that is still
+// bootstrapping. We put a node in the bootstrapping phase by giving it a
+// host-list that is impossible to join. When it is in this phase, we trigger
+// a heal attempt from another node. The heal should succeed from b's
+// perspective. Healing from bootstrap is very similar to joining a node that
+// is in the bootstrapping phase.
+func TestHealBeforeBootstrap(t *testing.T) {
+	a := newChannelNode(t)
+	defer a.Destroy()
+
+	b := newChannelNode(t)
+	defer b.Destroy()
+
+	// block a from completing bootstrap
+	block := make(chan struct{})
+	a.node.RegisterListener(on(JoinTriesUpdateEvent{}, func() {
+		<-block
+	}))
+
+	discoProvider := statichosts.New(a.node.Address(), b.node.Address())
+
+	// bootstrap a and wait for it to be in bootstrap phase
+	go func() {
+		_, _ = a.node.Bootstrap(&BootstrapOptions{
+			DiscoverProvider: discoProvider,
+		})
+	}()
+
+	// bootstrap b then add a to it's bootstrap provider after
+	bootstrapNodes(t, b)
+	b.node.discoverProvider = discoProvider
+	b.node.disseminator.ClearChanges()
+
+	// check that a is not yet part of b's membership
+	_, has := b.node.memberlist.Member(a.node.Address())
+	assert.False(t, has, "expected that a is not yet part of b's membership")
+
+	// start heal
+	targets := b.node.healer.Heal()
+	assert.Len(t, targets, 1, "expected that b healed with node node a")
+	assert.Equal(t, a.node.Address(), targets[0], "expected b healed with node a")
+
+	// a is now part of b's membership
+	_, has = b.node.memberlist.Member(a.node.Address())
+	assert.True(t, has, "expected that a is now part of b's membership")
+
+	// make sure that a is still not bootstrapped, this causes a to be suspect to be
+	ExecuteThenWaitFor(func() {
+		b.node.pingNextMember()
+	}, a.node, RequestBeforeReadyEvent{})
+
+	// allows a to bootstrap
+	ExecuteThenWaitFor(func() {
+		close(block)
+	}, a.node, JoinCompleteEvent{})
+
+	// progress timers so that incarnation numbers can bump
+	c := clock.NewMock()
+	c.Add(time.Millisecond)
+	a.node.clock = c
+
+	waitForConvergence(t, time.Second*3, a, b)
+}
+
+type partition []*testNode
+
+func newPartition(t *testing.T, n int) partition {
+	p := partition(genChannelNodes(t, n))
+	bootstrapNodes(t, p...)
+	waitForConvergence(t, 500*time.Millisecond, p...)
+	return p
+}
+
+func (p partition) AddPartitionWithStatus(B partition, status string) {
```

Comment:
Please document this function. E.g. "adds B (with status `status`) to the membership of p, and clears the disseminator".